### DISCUSSION TITLE
Issue when used as submodule

### DISCUSSION THREAD

**Discussion by TheLongRunSmoke** [bxparks/AUnit#79] [General]
Hi. For some reasons, I can't use AUnit as git submodule. When installed using Arduino IDE all works fine.

In my setup I got
```
In file included from ExternalTests.cpp:12:0:
AUnitMetaTest.h:13:12: fatal error: AUnit.h: No such file or directory
compilation terminated.
exit status 1
Error compiling for board Arduino Pro or Pro Mini.
```

**Steps to reproduce:**
1. Create sketch in empty folder.
2. Init git repo.
3. Add AUnit as submodule. 
```
 git submodule add git@github.com:bxparks/AUnit.git
```
4. Add minimal testing code in to the sketch.
```
#include "AUnit/src/AUnit.h"

test(submodule_include_test){
  int x = 1;
  assertEqual(x, 1);
}
```
5. Try to compile and get various errors.

**Comment by bxparks** [comment]
I don't use submodules. But if I were to guess, it's because your directory layout does not match the expectations of the Arduino IDE. The IDE is quite particular about where its libraries must live. The user-defined ones must live under `{sketchbook}/libraries/`, where `{sketchbook}` is the directory defined in the Preferences menu of the IDE. If you don't have the AUnit directory in the `./libraries/` directory, then things won't work. You can move the AUnit directory under there, or use a symlink under Unix/Linux. I don't use Windows for development, so I don't know if Windows shortcuts will work.

**Comment by TheLongRunSmoke** [comment]
Have same issue even if place it in ```libraries/```.
However, when I remove ```AUnit/tests``` folder it's start to compile and work normally, from any path. 
Looks like you unintentionally use includes with diamond brackets in tests. 

This need to be fixed. Libraries outside of source folder is generally a bad idea, due possibilities of version conflicts between projects.

**Comment by bxparks** [comment]
I'm not sure I understand what the problem is, what your propose fix is. I think it would be helpful if you included more info about your environment, your source directory structure, and the tools that you using.

You say that AUnit works when installed using the Arduino IDE. That is the installation process which is supported. I don't use git submodules, I have never tested it, I don't mention it in my docs, so I don't what your source tree looks like, so I have no idea what is broken. I normally use diamond brackets very intentionally.

**Comment by TheLongRunSmoke** [comment]
```
├── src                 # Source files.
│   ├── libs              
│   │   ├── 3rd         # 3rd-party libs as submodules.
│   │   │   ├── AUnit           
│   │   │   └── ...            
│   │   └── ...         # Other libs as files.
│   ├── main.cpp        # Entry point.
│   └── ...
└── project.ino         # Empty Arduino project file.
```
Other libraries as [U8g2](https://github.com/olikraus/U8g2_Arduino) of [FreeRTOS port](https://github.com/feilipu/Arduino_FreeRTOS_Library) work ok.

Arduino IDE do not have any packet manager, so I want to use C approach to ship right library versions.

**Comment by bxparks** [comment]
I agree that your directory structure is better in many ways, and I even attempted to use something like when I first started working with Arduino. But Arduino does not support it, and it just broke too many things. I find it too difficult and time consuming to fight Arduino's design decisions, so my suggestion is to just go with the Arduino Library Manager defaults.

**Comment by TheLongRunSmoke** [comment]
Ok. For now I will use a fork without ```tests``` folder.

**Comment by bxparks** [comment]
You might want to look at PlatformIO which uses a directory structure similar to yours. Third party libraries like AUnit should work. As a publisher of 3rd party libraries, I cannot use PlatformIO, but as a consumer, it should work just fine.

**Comment by bxparks** [comment]
Closing this, but moving to Discussions for future reference.
